// @vscode/prompt-tsx/dist/base/tokenizer/tokenizer.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation and GitHub. All rights reserved.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnyTokenizer = void 0;
const openai_1 = require("../openai");
class AnyTokenizer {
    constructor(countTokens) {
        this.countTokens = countTokens;
    }
    async tokenLength(text, token) {
        return this.countTokens(text, token);
    }
    async countMessageTokens(message) {
        return this.countTokens({
            role: this.toChatRole(message.role),
            content: message.content,
            name: message.name
        });
    }
    toChatRole(role) {
        switch (role) {
            case openai_1.ChatRole.User: return 1;
            case openai_1.ChatRole.Assistant: return 2;
            case openai_1.ChatRole.System: return 1;
            case openai_1.ChatRole.Function: return 1;
        }
    }
}
exports.AnyTokenizer = AnyTokenizer;
